src/main/java/org/dataloader/DataLoaderHelper.java

```diff
@@ -98,34 +99,74 @@ Object getCallContext() {
+        boolean cachingEnabled = loaderOptions.cachingEnabled();
+
+        if (!cachingEnabled) {
+            return CompletableFuture.completedFuture(Optional.empty());
+        }
+
+        if (!futureCache.containsKey(key)) {
+            return CompletableFuture.completedFuture(Optional.empty());
+        }
+
+        return futureCache.get(key).handleAsync((v, exception) -> {
+            if (v != null) {
+                stats.incrementCacheHitCount();
+                return Optional.of(v);
+            }
+
+            return Optional.empty();
+        });
+    }
+
+    CompletableFuture<V> loadFromSource(K key, Object loadContext) {
+        boolean batchingEnabled = loaderOptions.batchingEnabled();
+
+        CompletableFuture<V> future = new CompletableFuture<>();
+        if (batchingEnabled) {
+            loaderQueue.add(new LoaderQueueEntry<>(key, future, loadContext));
+        } else {
+            stats.incrementBatchLoadCountBy(1);
+            // immediate execution of batch function
+            future = invokeLoaderImmediately(key, loadContext);
+        }
+
+        return future;
+    }
 
     CompletableFuture<V> load(K key, Object loadContext) {
         synchronized (dataLoader) {
             Object cacheKey = getCacheKey(nonNull(key));
             stats.incrementLoadCount();
 
-            boolean batchingEnabled = loaderOptions.batchingEnabled();
             boolean cachingEnabled = loaderOptions.cachingEnabled();
 
-            if (cachingEnabled) {
-                if (futureCache.containsKey(cacheKey)) {
-                    stats.incrementCacheHitCount();
-                    return futureCache.get(cacheKey);
+            return loadFromCache(key).thenApplyAsync((cachedValue) -> {
+                if (cachedValue.isPresent()) {
+                    return cachedValue.get();
                 }
-            }
 
-            CompletableFuture<V> future = new CompletableFuture<>();
-            if (batchingEnabled) {
-                loaderQueue.add(new LoaderQueueEntry<>(key, future, loadContext));
-            } else {
-                stats.incrementBatchLoadCountBy(1);
-                // immediate execution of batch function
-                future = invokeLoaderImmediately(key, loadContext);
-            }
-            if (cachingEnabled) {
-                futureCache.set(cacheKey, future);
-            }
-            return future;
+                var sourceResult = loadFromSource(key, loadContext);
+                sourceResult.thenAcceptAsync((sourceValue) -> {
+                    if (cachingEnabled) {
+                        futureCache.set(cacheKey, java.util.concurrent.CompletableFuture.supplyAsync(() -> sourceValue));
+                    }
+                });
+
+                try {
+                    return sourceResult.get();
+                } catch (InterruptedException | java.util.concurrent.ExecutionException e) {
+                    return null;
+                }
```

Comment:
This is a blocking call - `sourceResult.get();`

DataLoader MUST always be async - ALWAYS.  We cant have this.